packages/next/src/server/app-render/segment-value-encoding.ts

```diff
@@ -0,0 +1,60 @@
+import { UNDERSCORE_NOT_FOUND_ROUTE } from '../../api/constants'
+import type { Segment as FlightRouterStateSegment } from './types'
+
+// TypeScript trick to simulate opaque types, like in Flow.
+type Opaque<K, T> = T & { __brand: K }
+
+export type EncodedSegment = Opaque<'EncodedSegment', string>
+
+export function encodeSegment(
+  segment: FlightRouterStateSegment
+): EncodedSegment {
+  if (typeof segment === 'string') {
+    const safeName =
+      segment === UNDERSCORE_NOT_FOUND_ROUTE
+        ? // TODO: FlightRouterState encodes Not Found routes as "/_not-found".
+          // But params typically don't include the leading slash. We should use
+          // a different encoding to avoid this special case.
+          '_not-found'
+        : encodeToFilesystemAndURLSafeString(segment)
+    // Since this is not a dynamic segment, it's fully encoded. It does not
+    // need to be "hydrated" with a param value.
+    return safeName as EncodedSegment
+  }
+  const name = segment[0]
+  const paramValue = segment[1]
+  const paramType = segment[2]
+  const safeName = encodeToFilesystemAndURLSafeString(name)
+  const safeValue = encodeToFilesystemAndURLSafeString(paramValue)
+
+  const encodedName = '$' + paramType + '$' + safeName + '$' + safeValue
```

Comment:
I changed the encoding a bit here in anticipation of our upcoming project to remove the dynamic params from the server response and instead extract them from the URL on the client. The server will send everything except for that final `safeValue`. And then the client will append it immediately when processing the payload.

Reply:
There's a comment about that here: https://github.com/vercel/next.js/pull/73853/files#diff-f296fd6b5d4b377330673eab8efc9709ba5dfbcde46a1dffa1d76adf400717c8R506-R509